クエリログや引数で与えたクエリから不正なクエリを検出するコマンドを追加

Komei22 opened the pull request:
### 何をするのか
引数で与えたクエリの文字列、またはクエリのログファイルを入力として与え，ブラックリストorホワイトリストベースで不正なクエリを検出するコマンドを追加します。

### 実行例
クエリの文字列から検出
```
❯ sqd check "DROP TABLE article" -m whitelist -l whitelist
Suspicious querys:
DROP TABLE article
```

クエリのログファイルから検出
```
❯ sqd checkl -q query.log -m whitelist-l whitelist
Suspicious querys:
SELECT * FROM articles
DROP TABLE articles
```

Commit:
コマンドラインツール化

Commit:
listをもとに不正クエリを検知するdetectorパッケージを追加

Commit:
クエリログから不正クエリを検知するコマンドを追加

fluentdなどでクエリのログを受け取り、受け取ったクエリログに不正クエリが含まれていないかを確認するようなユースケースを想定

Commit:
detectorのmodeをコマンドオプションとして指定できるように変更

読み込むリスト名で判別するよりもmodeとして指定したほうが良いと考えたため

Commit:
クエリ毎に改行して表示するように変更

Commit:
ファイル読み込みと配列への保存を関数で分離

Commit:
クエリをパースしてクエリ構造に変換する処理を追加

Commit:
不正クエリの検知と出力を関数で分離

一つの関数で2つのことをしていたので

Commit:
detectorのテストを追加

テストデータ登録のためにmatcherのメソッドの一部をexportに変更

Commit:
結果の表示する処理をcheckコマンド側に移動

detectorは不正なクエリのリストを返すところまでやり、結果の表示は上位のプログラムに任せる

Commit:
lintをかけた

Commit:
detectorのmodeをwhite/blacklistに限定

予期せぬ入力の受付をなくすため

Commit:
detectorのモード変更の処理を変更し、エラー処理を追加

予期せぬモードが入力された場合に対応

Commit:
detectorのテストを修正

Commit:
READMEをアップデート

Commit:
detectorの初期化を文字列とファイルからできるように変更

標準入力から入力されたクエリからも初期化できるようにするため

Commit:
標準入力で与えられたクエリが正当化を判定するコマンドを追加

Commit:
detectorのテストを修正

標準入出力に対応。エラー判定を修正

Commit:
不要な処理があったので削除

Commit:
detectorのテストメソッド名を修正

Commit:
エラーメッセージを修正

Commit:
タイポを直した

Commit:
readmeをアップデート

monochromegane commented on cmd/check.go:
```diff
@@ -0,0 +1,68 @@
+package cmd
```
最も気になったのは checkとchecklというほぼ処理内容の重複するサブコマンドが存在することです．
違いは候補となるクエリを直接受け取るかファイルから読み込むかだと思いますが，この実装だと例えば標準入力から受け付けるとなった時にまたサブコマンドを作ることになりそうです．
このような重複は同じ修正内容を複数箇所行わなければならないことによる修正漏れが発生するだけでなく，多人数でのメンテナンスになった際に際限なくコピペが繰り返されてしまう元となります．
できるだけこのような実装は回避しましょう．
具体的には クエリファイルをfオプション，直接クエリを指定する場合にqオプションなどを使ってもらうようにすれば解決すると思います．
そしてサブコマンドがcheck以外ないのであれば，sqd コマンドだけの方がシンプルですね．

monochromegane commented on README.md:
````diff
@@ -1,2 +1,31 @@
+
+```
+$ sqd check -m whitelist -q query.log -l whitelist
+Suspicious querys
````
queriesが正しい複数形ですね

monochromegane commented on README.md:
```diff
@@ -1,2 +1,31 @@
 suspicious query detection
+
+## Usage
+`sqd check` detect illegal query in query log file using list file(blacklist or whitelist).
```
illegalなのかsuspiciousなのか統一した方が良さそうです

monochromegane commented on cmd/check.go:
```diff
@@ -0,0 +1,68 @@
+
+func init() {
+	checkCmd.Flags().StringVarP(&listFilepath, "list", "l", "", "file path of blacklist or whitelist")
+	checkCmd.Flags().StringVarP(&detectorMode, "mode", "m", "whitelist", "detection mode (whitelist or blacklist)")
```
ブラックリスト/ホワイトリスト以外のパターンが存在しないと思うので -B, -W とかでどうでしょうか

Komei22 replied:
fc95a2e
`-B/-W listfile`と指定するようにして対応しました

monochromegane commented on cmd/check.go:
```diff
@@ -0,0 +1,68 @@
+
+		suspiciousQueries, err := d.Detect(m)
+		if err != nil {
+			fmt.Printf("Can't detection suspicious query. (%s)", err)
```
この辺りのコマンドラインツールの出力の扱いはcobraのテンプレートだとは思いますが，Goらしい，かつテストしやすいコマンドラインツールを作る上では https://deeeet.com/writing/2014/12/18/golang-cli-test/ この辺りのエントリをざっと目を通しておくと良いでしょう．余裕があれば取り入れてみてください．
cobraならOutput指定などを行った上でcmd.Printfなどで書けたはずです．

monochromegane commented on detector/detector.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+// New detector
+func New(i interface{}, mode Mode) (*Detector, error) {
```
型付けによって実装時のエラーを減らしていく方針のGo言語では型チェックの恩恵を受けることができない空インターフェースの使用は避けた方が良いです．

monochromegane commented on detector/detector.go:
```diff
@@ -0,0 +1,89 @@
+	return d, err
+}
+
+func (d *Detector) readFile(filepath string) error {
```
ここでクエリファイルから読み込んだクエリ一覧を一旦メモリに保存していますが，これは無駄だと思います．
全体を保持しなくても1行ずつ`isSuspiciousQuery`を適用することができるためです．
これは直接クエリを指定する場合でも同様のことが言えます．

monochromegane commented on detector/detector.go:
```diff
@@ -0,0 +1,89 @@
+}
+
+// Detect find suspicious queries using matcher(whitelist or blacklist based)
+func (d *Detector) Detect(m *matcher.Matcher) ([]string, error) {
```
Detectごとにmatcherを変更する必要がなければ都度 matcherを受け渡す必要はないと思います．
Newの時点で渡してはどうでしょうか．

monochromegane commented on detector/detector.go:
```diff
@@ -0,0 +1,89 @@
+		}
+	}
+	return suspiciousQueries, nil
+}
```
まとめるとNew(matcher, B/W)とDetect(query)で良さそうです．
複数クエリと単体クエリの処理共通化についてはfをqに寄せるのではなくqをfに寄せましょう．
つまり単体クエリを処理する部分を作り，ファイルから読み込む場合はそれを各行で呼ぶ．単体の場合はそれを直接呼ぶ，もしくは1行だけを読み込んだというように．

monochromegane commented on matcher/matcher.go:
```diff
@@ -33,10 +34,11 @@ func (m *Matcher) loadList(filepath string) error {
 
-func (m *Matcher) saveList(r io.Reader) error {
+// SaveList save queries to list
+func (m *Matcher) SaveList(r io.Reader) error {
```
saveだとファイルに保存する処理を想起してしまうので，readListとかで良い気がして来ました

monochromegane commented:
@Komei22 ざっとみました．修正内容が多めだと思うので修正おわったら教えてください．
改めてみます

Commit:
readmeをアップデート

Commit:
ブラックリストを-B、ホワイトリストを-Wオプションで検知のモードを指定できるように変更

Commit:
detectorを修正

- 入力されたクエリはメモリ上に保持する必要がないため保持しないように変更
- detectorの初期化時にmatcherをセットするように変更
- interface{}型を利用しないようにするために、クエリの文字列orファイル入力によって利用する関数を分けた

Commit:
checkとchecklのサブコマンドの処理をcheckに集約

Commit:
Detectを使って処理を共通化

Commit:
-lオプションを削除して、-B,=Wでそれぞれブラックリストとホワイトリストのファイルを指定するようにした

検知のモードとリストファイルは常に一致するため

Commit:
メソッド名を変更

Commit:
checkのサブコマンドをrootコマンドにお引越し

Commit:
matcherとdetectorのテストを修正

Commit:
rootCmdを関数で生成して実行するようにし、エラー出力をreturnするようにした

テストをやりやすくするために

Commit:
rootCmdのテストを追加

Komei22 commented:
@monochromegane レビューお願いします！

monochromegane commented on README.md:
````diff
@@ -1,2 +1,31 @@
+`sqd check` detect suspicious query in query log file using list file(blacklist or whitelist).
+
+```
+$ sqd check -m whitelist -q query.log -l whitelist
````
せっかくなのでREADMEも追従していきましょう

monochromegane commented on cmd/root.go:
```diff
@@ -0,0 +1,86 @@
+	}
+	rootCmd.Flags().StringVarP(&query, "query", "q", "", "query string")
+	rootCmd.Flags().StringVarP(&querylog, "file", "f", "", "query log file path")
+	rootCmd.Flags().StringVarP(&whitelist, "Whitelist", "W", "", "whitelist file path")
```
`BoolVarP` の方が適切ではないでしょうか

Komei22 replied:
ここはもともと`-l`でホワイト/ブラックリストを指定して、`-B`or`-W`で検知のモードを指定する仕様にしていたのですが、指定するリストと検知のモードは一致するので、`-B [blacklist file]` or `-W [whitelist file]`で良いと思いました。

monochromegane replied:
はあい

monochromegane commented on detector/detector.go:
```diff
@@ -0,0 +1,84 @@
+	}
+	defer r.Close()
+
+	var suspiciousQueries []string
```
detectFrom などで io.Readerを受けるメソッドを作ってそれでテストしていたら，アドバイス吸収しているなあ，ふむふむ〜と思えたなあ

Komei22 replied:
io.Readerで受けるメソッドにしていたなら、detectorのテストで利用できたなあ

これに関連して質問があります。
io.Readerで受けるメソッドを定義しておくことで、テスト時にはファイルを読み込まなくてもStringのio.Readerをメソッドに渡すことができ、テストケースをテストファイル内で管理できるところが良いところだと思っています。
しかし、[root_test.go](https://github.com/Komei22/sqd/pull/2/files#diff-c20737afa75dbf45a65e7088758c44e7R9)では、cliコマンドのテストを行うために、コマンド引数にファイルパスが必要なります。このような場合は、結局テストデータのファイルが必要になると思うので、テストデータをファイルとして管理した方が良いのでしょうか？

monochromegane replied:
そうですね．その利点が一つと，例えば標準入力を扱うようになった場合などにも提供する関数の使い方を変えずに済むといったメンテナンス性の向上も見込めますね．
僕の考えは，ファイルをオープンするという処理は標準ライブラリの十分にテストされた部分であることからファイルパスを引数に渡すという部分という部分のみのテストよりは（自分で実装した）中身をどう処理するかといったテストがあれば良いというものです．
そのため，上記のインターフェースを使ったやり方での実装をオススメしました．

「手元で確認するのがやりやすいのでファイル作っておきます」というのであればそれはそれで決めの問題なのでいいと思います．
この方針であれば，OSSを前提とすると，その（謎）ファイルを使った動作確認をする手順などが他の人がわかる形になっているといいですね（例えばREADMEの例でそれを使うとどのような結果が返って来るのかなど）

Komei22 replied:
なるほど。納得しました！
root_test.goのテストは標準ライブラリで十分にテストされているとみなして削除し、自分で書いたdetectorをテストすることで十分と判断しようと思いました。
また、DetectFromをio.Readerを受けるメソッドとして定義して、こちらでテストを書き直したいと思いました。

monochromegane reviewed (commented):
概ね良さそうです．コメントしましたが，以前アドバイスした内容と同じものを適用する機会もあったのでやって見てはどうかなと思います．（やらなくても大丈夫です）

Commit:
readmeをアップデート

Commit:
rootCmdのテストを削除

rootCmdは利用しているライブラリcobraでテストされている内容で十分と判断したため

Commit:
DetectFromの引数をio.Readerに変更

io.Readerにすることで今後標準出力などに対応するときに関数の使い方を変えなくて良くするため

Commit:
DetectFromメソッドでテストをするように変更

Commit:
実行結果を更新

Komei22 commented:
@monochromegane 
以下の変更を行いました。
- rootCmdのテストを削除
- DetectFromの引数をio.Readerで受けれるように変更（テストも更新）
- READMEを更新

Komei22 commented:
こちらコメント内容に対応したのでマージします〜
レビューありがとうございました！
